agent/lib/kontena/workers/log_worker.rb

```diff
@@ -51,31 +65,53 @@ def queue_processing?
 
     def start
       sleep 1 until Actor[:etcd_launcher].running?
+      start_container_workers
+      async.process_queue
+    end
+
+    def start_container_workers
       Docker::Container.all.each do |container|
         begin
           self.stream_container_logs(container) unless container.skip_logs?
         rescue Docker::Error::NotFoundError
           # Could be thrown since container.skip_logs? actually loads the container details
         end
       end
-
-      async.process_queue
     end
 
     def process_queue
+      current_actor = Actor.current
       defer {
-        while queue_processing? && data = @queue.pop
-          rpc_client.async.notification('/containers/log', [data])
-          if @queue.size > 100
-            sleep 0.001
-          else
-            sleep 0.05
-          end
+        queue = current_actor.queue
+        while current_actor.queue_processing? && data = queue.pop
+          process_queue_item(current_actor, data, (queue.size + 1))
         end
       }
     end
 
-    def stop
+    # @param [LogWorker] worker
+    # @param [Hash] data
+    # @param [Integer] queue_size
+    def process_queue_item(worker, data, queue_size)
+      rpc_client.async.notification('/containers/log', [data])
+      if queue_size > QUEUE_LIMIT && !worker.throttling?
+        warn "queue size is over #{QUEUE_LIMIT}, throttling logs until queue is processed"
+        worker.throttle!
+        worker.stop_container_workers
+      elsif queue_size > 100 && worker.throttling?
+        sleep 0.05
+      elsif queue_size > 100
+        sleep 0.01
+      elsif worker.throttling?
+        info "queue has been almost processed, releasing throttle"
+        worker.release_throttle!
+        worker.start_container_workers
+      else
+        sleep 0.05
+      end
+    end
+
+    def stop_container_workers
```

Comment:
The way this marks the current `Time.now` timestamp into etcd instead of the actual last processed log item timestamp means that we're going to lose any logs that are buffered in the Docker API queues.